.github/workflows/build_shadow.yml

```diff
@@ -49,44 +56,66 @@ jobs:
           # prevent prompts for user input during package install
           DEBIAN_FRONTEND: 'noninteractive'
         run: |
+          if [ "${{ matrix.cc }}" = "gcc" ]; then COMPILERS="gcc g++"
+          elif [ "${{ matrix.cc}}" = "clang" ]; then COMPILERS="clang"
+          else exit 1
+          fi
```

Comment:
This was a bit a bit awkward to handle in the matrix, and in general it'd be nice not to clutter the matrix dimensions with purely "derived" dimensions.

Ended up finding another way to dedupe this logic, though: basically created a dedicated step for figuring out these parameters and re-exporting them for subsequent steps.

Reply:
> This was a bit a bit awkward to handle in the matrix, and in general it'd be nice not to clutter the matrix dimensions with purely "derived" dimensions.

Actually I just noticed the latter point is moot; the GH UI doesn't include dimensions that are only set via the 'include' stanza, such as package_manager.

I'll take one more quick look to see if I can just fold these in there; when I tried it before GH ended up creating points in the matrix instead of supplementing the existing ones, but maybe I made a mistake...

Reply:
I *think* what's happening is that the GH CI will only use the first matching item in the 'include' list, so we can't easily use it to set multiple, orthogonal, derived dimensions. (e.g. package_manager based on container, and cxx based on cc).

It might be nice to move package_manager out of there for that reason, and either set an environment variable or have the dependency-install steps be conditional on the container directly. I'm going to leave that for a different PR, though.

Reply:
Yeah the includes feature isn't well documented. My understanding is that if only one item doesn't match, it will add that item to the existing matrix element, but if there are >1 items that don't match, it will create a new matrix element with those items.